File: Android/app/src/main/java/com/jiny/catchtherule/data/RankingService.kt
package com.jiny.catchtherule.data

import android.content.Context
import com.jiny.catchtherule.core.model.GameMode
import com.jiny.catchtherule.core.model.RankEntry

/**
 * 랭킹 백엔드 추상화. 실서버는 [MinigameRankingService], 오프라인 개발용은 [MockRankingService].
 */
interface RankingService {
    /** 점수 제출 후 내 순위를 반환. */
    suspend fun submit(score: Int, nickname: String, mode: GameMode): Int
    /** 리더보드 조회. */
    suspend fun leaderboard(mode: GameMode): List<RankEntry>
}

/**
 * 앱 전역 랭킹 서비스. [init] 으로 한 번 실서버 구현을 주입한다(MainActivity 에서 호출).
 * 미초기화 시에는 안전하게 Mock 으로 폴백.
 */
object Ranking {
    @Volatile
    private var injected: RankingService? = null

    val service: RankingService
        get() = injected ?: MockRankingService

    fun init(context: Context) {
        if (injected == null) {
            injected = MinigameRankingService(context.applicationContext)
        }
    }
}

File: iOS/CatchTheRule/Services/RankingService.swift
import Foundation

/// 랭킹 백엔드 추상화. 실제 Minigame 서버 스펙이 확정되면
/// `MinigameRankingService` 를 구현해 교체한다. 그 전까지는 Mock 사용.
protocol RankingService {
    /// 점수 제출 후 내 순위를 반환.
    func submit(score: Int, nickname: String, mode: GameMode) async throws -> Int
    /// 리더보드 조회.
    func leaderboard(mode: GameMode) async throws -> [RankEntry]
}

enum RankingError: Error {
    case notConfigured
    case server(String)
}

/// 앱 전역에서 사용할 현재 랭킹 서비스.
/// 실서버: MinigameRankingService() · 오프라인 개발용: MockRankingService.shared
enum Ranking {
    static let service: RankingService = MinigameRankingService()
}
